File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
n, m = map(int,input().split())
trees = sorted(list(map(int, input().split())))

'''
4 7
20 15 10 17
'''
start = 1
end = sum(trees)
while start <= end:
    mid = (start + end) // 2
    cut = sum(i - mid for i in trees if i > mid)

    # 자른게 더 많다 == cut 이 너무 낮다 == mid 를 높여야한다
    if cut >= m:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;

public class Main {
  static long cut(int target, int[] arr){
    long cnt = 0;
    for (int i = 0; i < arr.length; i++) {
      if (arr[i] > target) cnt += (arr[i] - target);
    }
    return cnt;
  }

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int[] inputs = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
    int N = inputs[0];
    int M = inputs[1];
    int[] arr =  Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

    Arrays.sort(arr);
    int s = 1;
    int e = arr[N-1];
    int ans = 0;
    while (s <= e){
      int mid = (s +e)/2;
      if (cut(mid, arr) >= M) {
        s = mid + 1;
        ans = mid;
      }
      else e = mid - 1;
    }
    // why
    System.out.println(ans);

  }
}
